lib/Parse/ParseDecl.cpp

```diff
@@ -2005,13 +2005,28 @@ ParserStatus Parser::parseDecl(SmallVectorImpl<Decl*> &Entries,
                                                   false);
         }
       }
+      if (CurDeclContext) {
+        auto ty =
+          CurDeclContext->getAsNominalTypeOrNominalTypeExtensionContext();
+        if (ty) {
+          diagnose(ty->getLoc(), diag::note_in_decl_extension, false,
+                   ty->getNameStr());
+        } else if (auto ed = dyn_cast<ExtensionDecl>(CurDeclContext)) {
+          if (auto extension = ed->getExtendedType()) {
+            diagnose(ed->getLoc(), diag::note_in_decl_extension, true,
+                     StringRef(extension->getString()));
+          } else {
+            diagnose(ed->getLoc(), diag::note_in_extension);
```

Comment:
`getAsNominalTypeOrNominalTypeExtensionContext()` would not return a valid type for extensions (`base=<null>` from dumping the context). My guess is this happens during the first pass?